packages/crayons-core/src/components/form/form.tsx

```diff
@@ -0,0 +1,422 @@
+
+    const validationErrors = await this.handleValidation();
+
+    const keys = [
+      ...Object.keys(this.values),
+      ...Object.keys(validationErrors),
+    ];
+
+    keys.forEach(
+      (k: string) => (touchedState = { ...touchedState, [k]: true })
+    );
+    // on clicking submit, mark all fields as touched
+    this.touched = { ...this.touched, ...touchedState };
+
+    isValid = !validationErrors || Object.keys(validationErrors).length === 0;
+
+    console.log({ values: this.values, errors: this.errors, isValid });
+
+    this.isSubmitting = false;
+
+    return { values: this.values, errors: this.errors, isValid };
+  };
+
+  handleReset = async (event?: Event): Promise<void> => {
+    event?.preventDefault();
+    event?.stopPropagation();
+    this.isSubmitting = false;
+    this.values = this.formInitialValues;
+
+    let touchedState = {};
+    const initialValuesKeys = Object.keys(this.initialValues);
+
+    initialValuesKeys.forEach(
+      (k: string) => (touchedState = { ...touchedState, [k]: true })
+    );
+    this.touched = touchedState;
+
+    if (initialValuesKeys && initialValuesKeys.length > 0) {
+      this.errors = await this.handleValidation();
+    }
+
+    this.focused = null;
+  };
+
+  handleValidation = async () => {
+    this.isValidating = true;
+
+    let validationErrors = {};
+
+    // run validations against validationSchema if present
+    if (
+      this.formValidationSchema &&
+      Object.keys(this.formValidationSchema).length
+    ) {
+      const pr = validateYupSchema(
+        prepareDataForValidation(this.values),
+        this.formValidationSchema
+      );
+      try {
+        await pr;
+        validationErrors = {}; // reset errors if no errors from validation
+      } catch (err) {
+        validationErrors = yupToFormErrors(err);
+      }
+    }
+
+    // run validations with validate function if passed as prop and merge with the errors from the above step
+    if (this.validate && typeof this.validate === 'function') {
+      try {
+        validationErrors = {
+          ...validationErrors,
+          ...((await this.validate(this.values)) || {}),
+        };
+      } catch (err) {
+        console.error(`Error in calling validate function ${err.message}`);
+        validationErrors = { ...validationErrors };
+      }
+    }
+
+    this.errors = validationErrors;
+
+    this.isValidating = false;
+    return validationErrors;
+  };
+
+  handleInput = async (event: Event) => {
+    event.stopPropagation();
+    const { field, value } = (event as any).detail;
+    this.values = { ...this.values, [field]: value };
+
+    /** Validate, if user wants to validateOnInput */
+    if (this.validateOnInput) {
+      this.touched = { ...this.touched, [field]: true };
+      await this.handleValidation();
+    }
+  };
+
+  handleBlur = async (event: Event) => {
+    event.stopPropagation();
+    const { field, value } = (event as any).detail;
+    if (this.focused) this.focused = null;
```

Comment:
handleInput and handleBlur seem to be doing the same thing! what is the significance of focused?